src/tls/test/main.cpp

```diff
@@ -224,3 +249,85 @@ TEST_CASE("Manually hash, sign, verify, with certificate")
     CHECK_FALSE(verifier->verify(hash, signature));
   }
 }
+
+TEST_CASE("Recoverable signatures")
+{
+  auto kp = tls::KeyPair_k1Bitcoin(MBEDTLS_ECP_DP_SECP256K1);
+
+  vector<uint8_t> contents(contents_.begin(), contents_.end());
+  tls::HashBytes hash = bad_manual_hash(contents);
+
+  auto signature = kp.sign_recoverable_hashed(hash);
+  const auto target_pem = kp.public_key_pem().str();
+
+  auto recovered = tls::PublicKey_k1Bitcoin::recover_key(signature, hash);
+
+  {
+    INFO("Normal recovery");
+    CHECK(target_pem == recovered.public_key_pem().str());
+  }
+
+  // NB: Incorrect arguments _may_ cause the verification to throw with no
+  // recoverable key, but they may simply cause a different key to be returned.
+  // These tests look for either type of failure.
+
+  {
+    INFO("Corrupted hash");
+    auto hash2(hash);
+    corrupt(hash2);
+    bool recovery_failed = false;
+    try
+    {
+      auto r = tls::PublicKey_k1Bitcoin::recover_key(signature, hash2);
+      recovery_failed = target_pem != r.public_key_pem().str();
+    }
+    catch (const std::exception& e)
+    {
+      recovery_failed = true;
+    }
+    CHECK(recovery_failed);
```

Comment:
In these negative tests, it is not clear whether `recovery_failed` is set to `true` because an exception has been raised by `recover_key()` or because `target_pem != r.public_key_pem().str()`. Would it be worth tightening these test cases?

Reply:
Annoyingly we can't predict which failure path we'll hit. When corrupting the message hash or the signature, there seems to be about a 70% chance that we still get _a_ key recovered, just not the correct key. So for the test, we just want to confirm that corruption doesn't give the same key, in whatever form.